napari/utils/_tests/test_proxies.py

```diff
@@ -58,24 +60,58 @@ def __getitem__(self, key):
     assert proxy.x.method() == 2
 
     assert isinstance(proxy, Tester)
-    with pytest.warns(FutureWarning) as e:
+    with pytest.warns(FutureWarning, match='Private attribute access'):
         proxy._private
-    assert 'Private attribute access' in str(e[0].message)
 
-    with pytest.warns(FutureWarning) as e:
+    with pytest.warns(FutureWarning, match='Private attribute access'):
+        # warns on setattr
+        proxy._private = 4
+
+    with pytest.warns(FutureWarning, match='Private attribute access'):
         # works on sub-objects too
         proxy.x._b
-    assert 'Private attribute access' in str(e[0].message)
 
-    with pytest.warns(FutureWarning) as e:
+    with pytest.warns(FutureWarning, match='Private attribute access'):
         # works on sub-items too
         proxy[0]._b
-    assert 'Private attribute access' in str(e[0].message)
 
     assert '_private' not in dir(proxy)
     assert '_private' in dir(t)
 
 
+@pytest.mark.skipif("qtpy" not in sys.modules, reason="requires Qt")
+def test_thread_proxy_guard(monkeypatch, qapp):
+    class X:
+        a = 1
+
+    called = False
+
+    def set_a_attr_ex(o):
+        nonlocal called
+        with pytest.raises(RuntimeError):
+            o.a = 2
+        called = True
+
+    def set_a_attr(o):
+        o.a = 3
+
+    monkeypatch.setenv('NAPARI_ENSURE_PLUGIN_MAIN_THREAD', 'True')
+
+    x = X()
+    x_proxy = PublicOnlyProxy(x)
+
+    t = Thread(target=set_a_attr_ex, args=(x_proxy,))
```

Comment:
I'd recommend implementing this with a `ThreadPoolExecutor` fixture instead of using threads directly because the test implementation will be shorter and more readable (especially with respect to the exception). Something like the following.

```python
@pytest.mark.skipif("qtpy" not in sys.modules, reason="requires Qt")
def test_thread_proxy_guard(monkeypatch, qapp, single_threaded_executor):
    class X:
        a = 1

    monkeypatch.setenv('NAPARI_ENSURE_PLUGIN_MAIN_THREAD', 'True')

    x = X()
    x_proxy = PublicOnlyProxy(x)

    f = single_threaded_executor.submit(x.__setattr__, 'a', 2)
    f.result()
    assert x.a == 2

    f = single_threaded_executor.submit(x_proxy.__setattr__, 'a', 3)
    with pytest.raises(RuntimeError):
        f.result()
    assert x.a == 2
```

Reply:
done